utils/vptr/include/virtual_ptr.hpp

```diff
@@ -0,0 +1,510 @@
+/***************************************************************************
+ *  Copyright (C) 2017 Codeplay Software Limited
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  For your convenience, a copy of the License has been included in this
+ *  repository.
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  Codeplay's ComputeCpp SDK
+ *
+ *  virtual_ptr.hpp
+ *
+ *  Description:
+ *    Interface for SYCL buffers to behave as a non-deferrenciable pointer
+ *
+ * Authors:
+ *
+ *    Ruyman Reyes   Codeplay Software Ltd.
+ *    Mehdi Goli     Codeplay Software Ltd.
+ *    Vanya Yaneva   Codeplay Software Ltd.
+ *
+ **************************************************************************/
+
+#include <CL/sycl.hpp>
+#include <iostream>
+
+#include <queue>
+#include <unordered_map>
+
+#ifndef VIRTUAL_PTR_VERBOSE
+// Show extra information when allocating and de-allocating
+#define VIRTUAL_PTR_VERBOSE 0
+#endif  // VIRTUAL_PTR_VERBOSE
+
+namespace codeplay {
+
+using buffer_data_type = uint8_t;
+using sycl_acc_target = cl::sycl::access::target;
+using sycl_acc_mode = cl::sycl::access::mode;
+
+/**
+ * PointerMapper
+ *  Associates fake pointers with buffers.
+ *
+ */
+class PointerMapper {
+ public:
+  /* pointer information definitions
+   */
+  static const unsigned long ADDRESS_BITS = sizeof(void *) * 8;
```

Comment:
Actually, this is not needed in this class, so I have removed it entirely. c9fbe84303b714dc2f42d03e28dd62e7b3faaafa